### ISSUE TITLE
[Bug]: datasource reserved_instance_offering_data_source unusable in current state

### ISSUE THREAD

**Issue by wevanscfi** [hashicorp/terraform-provider-aws#28358]
### Terraform Core Version

1.1.9

### AWS Provider Version

4.46.0

### Affected Resource(s)

reserved_instance_offering_data_source

### Expected Behavior

```terraform
data "aws_rds_reserved_instance_offering" "test" {
  db_instance_class   = "db.r5.xlarge"
  duration            = 31536000
  multi_az            = false
  offering_type       = "No Upfront"
  product_description = "postgresql"
}
```

Should return a valid RDS instance offering. In cases where there are multiple results returned due to partial matches on product_description, the datasource should return the offering that is an exact match.

### Actual Behavior

However, this will always raise an error, as there is no way to uniquely select just the "postgresql" product description. It will always find both the "postgresql" and "aurora_postgresql" product_descriptions and therefore always error.

### Relevant Error/Panic Output Snippet

```shell
Error: More than one rds Reserved Instance Offering found matching criteria; try different search

https://github.com/hashicorp/terraform-provider-aws/blob/f2bd7eecca5363d2dd4dd75fc2e80afafa9df066/internal/service/rds/reserved_instance_offering_data_source.go#L87
```

```shell
$ aws rds describe-reserved-db-instances-offerings --product-description "postgresql" --offering-type "No Upfront" --db-instance-class "db.r5.xlarge" --no-multi-az
{
    "ReservedDBInstancesOfferings": [
        {
            "ReservedDBInstancesOfferingId": "8de62f46-db02-444c-b38f-e4aa31c77c82",
            "DBInstanceClass": "db.r5.xlarge",
            "Duration": 31536000,
            "FixedPrice": 0.0,
            "UsagePrice": 0.0,
            "CurrencyCode": "USD",
            "ProductDescription": "aurora-postgresql",
            "OfferingType": "No Upfront",
            "MultiAZ": false,
            "RecurringCharges": [
                {
                    "RecurringChargeAmount": 0.38,
                    "RecurringChargeFrequency": "Hourly"
                }
            ]
        },
        {
            "ReservedDBInstancesOfferingId": "96443c04-892b-4350-9b23-52866b9547d9",
            "DBInstanceClass": "db.r5.xlarge",
            "Duration": 31536000,
            "FixedPrice": 0.0,
            "UsagePrice": 0.0,
            "CurrencyCode": "USD",
            "ProductDescription": "postgresql",
            "OfferingType": "No Upfront",
            "MultiAZ": false,
            "RecurringCharges": [
                {
                    "RecurringChargeAmount": 0.289,
                    "RecurringChargeFrequency": "Hourly"
                }
            ]
        }
    ]
}
```

### Terraform Configuration Files

```terraform
data "aws_rds_reserved_instance_offering" "test" {
  db_instance_class   = "db.r5.xlarge"
  duration            = 31536000
  multi_az            = false
  offering_type       = "No Upfront"
  product_description = "postgresql"
}
```

### Steps to Reproduce

Attempt to utilize the aws_rds_reserved_instance_offering datasource for any instance_class that offers both "postgresql" and "aurora_postgresql" products. There is no way to return a single "postgresql" instance offering.

### Would you like to implement a fix?

None

**Comment by siavashs** [comment]
>the datasource should return the offering that is an exact match

Even with an exact match the datasource might return more than one offering, I think a better implementation would be to either:
- return the first exact match
- return a list and let the user choose using an index, which I think is the better option

I can submit a fix if the maintainers specify which approach they would agree on.

**Comment by seanturner026** [comment]
Another option would be to configure the `product_description` to accept specific values that are documented such as `rds_postgresql` and `aurora_postgresql` and filter the API response accordingly. This should keep the same interface and make it possible for the data object to only return one offering.

**Comment by idonca** [comment]
any news here?

**Comment by seanturner026** [comment]
I took a crack at it but never finished it. Probably pretty close tbh. 

https://github.com/hashicorp/terraform-provider-aws/compare/main...seanturner026:terraform-provider-aws:b-reserved_instance_offering_data_source-refactor

**Comment by renehernandez** [comment]
I ran into this issue recently, and I implemented a workaround using the following components:

- `external` data source https://registry.terraform.io/providers/hashicorp/external/latest/docs/data-sources/external
- a bash script to run aws cli commands

**external data source definition**

```hcl
data "external" "offering" {
  program = ["bash", "${path.module}/scripts/get_offering_id.sh"]

  query = {
    offering_id                = var.offering_id
    instance_class          = var.instance_class
    multi_az                     = var.multi_az
    offering_type            = var.offering_type
    product_description = var.product_description
    duration                     = var.duration
  }

  lifecycle {
    postcondition {
      condition     = length(keys(self.result)) > 0
      error_message = "No offering ID was returned"
    }

    # If result map is empty, we don't need to check anything else since the previous postcondition will fail
    postcondition {
      condition     = length(keys(self.result)) == 0 ? true : var.instance_class == self.result.instance_class
      error_message = "The instance class in the offering ID does not match the requested instance class. There may be a bug in the script or if the offering ID was specified it may be incorrect."
    }

    postcondition {
      condition     = length(keys(self.result)) == 0 ? true : var.multi_az == tobool(lower(self.result.multi_az))
      error_message = "The multi-AZ configuration in the offering ID does not match the requested multi-AZ configuration. There may be a bug in the script or if the offering ID was specified it may be incorrect."
    }

    postcondition {
      condition     = length(keys(self.result)) == 0 ? true : var.offering_type == self.result.offering_type
      error_message = "The offering type in the offering ID does not match the requested offering type. There may be a bug in the script or if the offering ID was specified it may be incorrect."
    }

    postcondition {
      condition     = length(keys(self.result)) == 0 ? true : var.product_description == self.result.product_description
      error_message = "The product description in the offering ID does not match the requested product description. There may be a bug in the script or if the offering ID was specified it may be incorrect."
    }

    postcondition {
      condition     = length(keys(self.result)) == 0 ? true : var.duration == tonumber(self.result.duration) / 31536000 # Convert seconds to years
      error_message = "The duration in the offering ID does not match the requested duration. There may be a bug in the script or if the offering ID was specified it may be incorrect."
    }
  }
}
```

**bash script:**

```shell
#!/usr/bin/env bash

# For testing locally, you can run this script with the following command:
# ./scripts/get_offering_id.sh
# {"duration":"1","instance_class":"db.t4g.micro","multi_az":"false","offering_type":"No Upfront","product_description":"mysql"}

# Read input from stdin
read -r input

echo "Input: $input" > log.out

# Parse input JSON
instance_class=$(echo "$input" | jq -r '.instance_class')
offering_type=$(echo "$input" | jq -r '.offering_type')
product_description=$(echo "$input" | jq -r '.product_description')
duration=$(echo "$input" | jq -r '.duration')
multi_az=$(echo "$input" | jq -r '.multi_az')
offering_id=$(echo "$input" | jq -r '.offering_id | select (.!=null)')

echo -e "\nParsed Input: instance_class=$instance_class, offering_type=$offering_type, product_description=$product_description, duration=$duration, multi_az=$multi_az, offering_id=$offering_id" >> log.out

# Convert multi_az to AWS CLI flag
if [[ "$multi_az" == "true" ]]; then
  multi_az_flag="--multi-az"
else
  multi_az_flag="--no-multi-az"
fi

if [[ "$offering_id" != "" ]]; then
  command=(aws rds describe-reserved-db-instances-offerings --reserved-db-instances-offering-id "$offering_id")
else
  command=(aws rds describe-reserved-db-instances-offerings \
  --db-instance-class "$instance_class" \
  --product-description "$product_description" \
  --duration "$duration" \
  $multi_az_flag)
fi

echo -e "\nCommand: ${command[*]}" >> log.out

# Run AWS CLI command to describe reserved DB instances offerings
offerings=$( "${command[@]}" )

echo -e "\nOfferings: $offerings" >> log.out

if [[ "$offerings" == "" ]]; then
  echo "No offerings found" >> log.out

  echo "{}"
  exit 0
fi

if [[ "$offering_id" != "" ]]; then
  # If an offering ID was provided, there is a single offering and we just extract it from the array
  offering=$(echo "$offerings" | jq -r '.ReservedDBInstancesOfferings[] ')
else
  # Extract the offering ID that matches the product description and offering type exactly
  # For some reason that I couldn't  figure out, the above aws command fails if I include the --offering-type flag, most likely due to quoting issues
  # So, I'm filtering the results here instead
  offering=$(echo "$offerings" | jq -r --arg product_description "$product_description" --arg offering_type "$offering_type" \
    '.ReservedDBInstancesOfferings[] | select(.ProductDescription == $product_description and .OfferingType == $offering_type)')
fi

echo -e "\nOffering: $offering" >> log.out

formatted_output=$(echo "$offering" | jq -c '{offering_id: .ReservedDBInstancesOfferingId, instance_class: .DBInstanceClass, multi_az: (.MultiAZ | tostring), offering_type: .OfferingType, product_description: .ProductDescription, duration: (.Duration | tostring)}')

echo -e "\nFormatted Output: $formatted_output" >> log.out

# Output the result as JSON
echo "$formatted_output"
```

**Comment by YakDriver** [comment]
Thanks for raising this and our apologies for the delay. This will be fixed with #40281.
